Associations

ahzia opened the pull request:
In this milestone I worked on:

- Create classes "Classroom, Book, Rental"
- Create the relationships

kindly review my code and share your feedback.

o-rashad commented on book.rb:
```diff
@@ -0,0 +1,10 @@
+class Book
+  attr_accessor :title, :author
+  attr_reader :rentals
+
```
A method for adding rentals should be present, and Inside this method the book itself should be sent as a parameter to create the new rental.

o-rashad commented on person.rb:
```diff
@@ -1,14 +1,15 @@
 class Person
   attr_accessor :name, :age
-  attr_reader :id
+  attr_reader :id, :rentals
```
Getters and Setters for `@rentals` should be present

ahzia replied:
I don't think we may need the setter method for rentals, because rentals are an array of Rentals objects, and whenever we create a new Rental to a specific book we will push a new item to the array.
but I Implemented it as you requested.

o-rashad commented on person.rb:
```diff
@@ -1,14 +1,15 @@
     @correct_instance = Corrector.new
+    @rentals = []
   end
 
```
A method for adding rentals should be present. Link to an example and inside the person itself should be sent as a parameter to create the new rental.

o-rashad commented on rental.rb:
```diff
@@ -0,0 +1,11 @@
+class Rental
+  attr_accessor :date
```
Getters and Setters for `@book` and `@person` instance variables should be present

od-c0d3r reviewed (changes requested):
# Status: Changes Required ♻️
> “Change is the only constant in life.”
>-- idk who said it 😸

Asslam Alikom @ahzia 👋😄,
I'm Omar and I will be your reviewer for this project 😊

You almost completed the whole project in one run 👍
1. Good work, you were able to create `has-many/belongs-to` relationship between `Classroom` and `Student` as specified in the project requirements . 💯 
1. Created a class `Classroom`, `Book`, `Rental`  with the specified instance variables with the corresponding setters/getters.
1. Following Ruby best practices. 👌😎 

However, there is still something we wanna talk about which is creating the `many-to-many` relationship between `Person` and `Book` using the intermediate class `Rental`. 

In order we accomplish this relation we need to work in three different files with four different relations (you already know how you to do it .. like in `Student<=>Classroom` relationship) as `Rental` will be our common class between the other two.

1. Book `has-many` Rentals:
   1. - [x] `@rentals` should be initialized. [You already did it! 👌]
   2. - [x] A method for adding rentals should be present. [Link to an example](https://github.com/microverseinc/curriculum-ruby/blob/main/oop/articles/oop_relationships_by_examples.md#many-to-many-relationship)
   3. - [x] Inside this method the book itself should be sent as a parameter to create the new rental.
2. Person `has-many` Rentals:
    1. - [ ] Getters and Setters for `@rentals` should be present.
    2. - [x] `@rentals` should be initialized.
    3. - [ ] A method for adding rentals should be present. [Link to an example](https://github.com/microverseinc/curriculum-ruby/blob/main/oop/articles/oop_relationships_by_examples.md#many-to-many-relationship)
    4. - [ ] The person itself should be sent as a parameter to create the new rental.
3. and 4. Rental `belongs-to` Person && Rental `belongs-to` Book :
    1. - [ ] Getters and Setters for `@book` and `@person` instance variables should be present
    2. - [x] The constructor of Rental should be modified so Book and Person are set in it
    3. - [x] The rental is adding itself to the `@book`
    4. - [x] The rental is adding itself to the `@person` 

## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

Happy coding! 👋:heart:

> Feel free to leave any questions or comments in the PR thread if something is not 100% clear.

> **_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

ahzia commented:
@od-c0d3r Alaikum Salam 👋
Thanks for reviewing my project, I implemented all required changes.

- [x] added the "rent" method in Person and Book

- [x] added the missing getters and setters

o-rashad commented:
Hey @ahzia, 👋

Your project still needs to be reviewed by another code reviewer..
I know your dashboard maybe says otherwise..

So please work on the next project till the dev team fix the issue,
and soon you gonna be able to request for another Code Review  

Sorry for inconvenience,

ahzia commented:
@od-c0d3r I already submitted my project, and it shows me "Completed" in the dashboard, I will wait until tomorrow if the status changed back to "Request Review", I will request one.
Thanks for the reply.

bellom commented:
Hi @ahzia,

**Great work making all the necessary corrections from the previous review.**

I did a re-review, everything looks good now.

## Status: Approved 👍 

_Good luck in the next milestone :slightly_smiling_face:_ 

**Best**
